File: app/src/main/java/io/github/golok56/manajemenuang/ui/user/newuser/NewUserDialog.java
package io.github.golok56.manajemenuang.ui.user.newuser;

import android.app.AlertDialog;
import android.app.Dialog;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.golok56.manajemenuang.R;
import io.github.golok56.manajemenuang.ui.base.BaseDialog;

/**
 * @author Satria Adi Putra
 */

public class NewUserDialog extends BaseDialog {
    private static final String OK = "OK";
    private static final String CANCEL = "Cancel";

    @BindView(R.id.et_user_name_dialog_create_user)
    EditText mEtName;
    @BindView(R.id.et_user_wallet_dialog_create_user)
    EditText mEtWallet;

    private NewUserViewModel mViewModel;

    private Communicator mCommunicator;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = ViewModelProviders.of(this).get(NewUserViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             Bundle savedInstanceState) {
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof Communicator) {
            mCommunicator = (Communicator) context;
        } else {
            throw new RuntimeException("Error! Activity must implement Communicator interface!");
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        View view = getActivity().getLayoutInflater().inflate(R.layout.dialog_new_user, null);
        ButterKnife.bind(this, view);

        return new AlertDialog.Builder(getActivity())
                .setView(view)
                .setPositiveButton(OK, (dialogInterface, i) -> {
                    mViewModel.createUser(
                            mEtName.getText().toString(),
                            mEtWallet.getText().toString()
                    );
                    mCommunicator.onPositive();
                })
                .setNegativeButton(CANCEL, (dialogInterface, i) -> {
                    showToast(mViewModel.getValidationText());
                    getActivity().finish();
                })
                .create();
    }

    public interface Communicator {
        void onPositive();
    }

}

File: app/src/main/java/io/github/golok56/manajemenuang/views/fragment/dialog/NewUserDialog.kt
package io.github.golok56.manajemenuang.views.fragment.dialog

import android.app.AlertDialog
import android.app.Dialog
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import io.github.golok56.manajemenuang.R
import io.github.golok56.manajemenuang.viewmodels.NewUserViewModel
import kotlinx.android.synthetic.main.dialog_new_user.*

/**
 * @author Satria Adi Putra
 */

class NewUserDialog : BaseDialog() {
    private var mViewModel: NewUserViewModel? = null

    private var mCommunicator: Communicator? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModel = ViewModelProviders.of(this).get(NewUserViewModel::class.java)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is Communicator) {
            mCommunicator = context
        } else {
            throw RuntimeException("Error! Activity must implement Communicator interface!")
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = activity!!.layoutInflater.inflate(R.layout.dialog_new_user, null)

        return AlertDialog.Builder(activity)
                .setView(view)
                .setPositiveButton(OK) { dialogInterface, i ->
                    mViewModel!!.createUser(
                            etUsername.text.toString(),
                            etWallet.text.toString()
                    )
                    mCommunicator!!.onPositive()
                }
                .setNegativeButton(CANCEL) { dialogInterface, i ->
                    showToast(mViewModel!!.validationText)
                    activity!!.finish()
                }
                .create()
    }

    interface Communicator {
        fun onPositive()
    }

    companion object {
        private val OK = "OK"
        private val CANCEL = "Cancel"
    }

}
